File: app/src/main/java/me/eigenein/nexttrainwear/Station.java
package me.eigenein.nexttrainwear;

public class Station {

    public final String code;
    public final String longName;
    public final String land;
    public final float latitude;
    public final float longitude;

    public Station(
        final String code,
        final String longName,
        final String land,
        final float latitude,
        final float longitude
    ) {
        this.code = code;
        this.longName = longName;
        this.land = land;
        this.latitude = latitude;
        this.longitude = longitude;
    }

    @Override
    public String toString() {
        return String.format("%s (%s): %s, %s", longName, code, latitude, longitude);
    }
}

File: app/src/main/java/me/eigenein/nexttrainwear/Station.kt
package me.eigenein.nexttrainwear

import android.location.Location

data class Station(
    val code: String,
    val longName: String,
    val land: String,
    val latitude: Double,
    val longitude: Double
) {

    fun distanceTo(latitude: Double, longitude: Double): Float {
        val distance = FloatArray(1)
        Location.distanceBetween(this.latitude, this.longitude, latitude, longitude, distance)
        return distance[0]
    }

    companion object {
        fun findNearestStation(latitude: Double, longitude: Double): Station? =
            Stations.STATIONS.minBy { it.distanceTo(latitude, longitude) }
        fun findNearestStation(location: Location): Station? =
            findNearestStation(location.latitude, location.longitude)
    }
}
